### ISSUE TITLE
Covariance check with `_assert_covariance` is insufficient, unsound

### ISSUE THREAD

**Issue by steffahn** [Voultapher/self_cell#49]
```rs
use std::{cell::RefCell, fmt};

use self_cell::self_cell;

self_cell! {
    struct WrongVarianceExample {
        owner: (),
        #[covariant]
        dependent: Dependent,
    }
}

// this type is not covariant
type Dependent<'a> = RefCell<Box<dyn fmt::Display + 'a>>;
```
`Dependent<'a>` is not covariant in `'a`, however unsize coercion still allows an *OWNED* value
```rs
RefCell<Box<dyn fmt::Display + 'x>>
```
to be coerced into
```rs
RefCell<Box<dyn fmt::Display + 'y>>
```
for lifetimes `'x: 'y`, hence `_assert_covariance` compiles fine.

Unsoundness follows.
```rs
fn main() {
    let cell = WrongVarianceExample::new((), |_| RefCell::new(Box::new("")));
    let s = String::from("Hello World");

    // borrow_dependent unsound due to incorrectly checked variance
    *cell.borrow_dependent().borrow_mut() = Box::new(s.as_str());

    // s still exists
    cell.with_dependent(|_, d| println!("{}", d.borrow()));

    drop(s);

    // s is gone
    cell.with_dependent(|_, d| println!("{}", d.borrow()));
}
```
([run in Rust Explorer](https://www.rustexplorer.com/b#LyoKW2RlcGVuZGVuY2llc10Kc2VsZl9jZWxsID0gIjEuMC4xIgoqLwoKdXNlIHN0ZDo6e2NlbGw6OlJlZkNlbGwsIGZtdH07Cgp1c2Ugc2VsZl9jZWxsOjpzZWxmX2NlbGw7CgpzZWxmX2NlbGwhIHsKICAgIHN0cnVjdCBXcm9uZ1ZhcmlhbmNlRXhhbXBsZSB7CiAgICAgICAgb3duZXI6ICgpLAogICAgICAgICNbY292YXJpYW50XQogICAgICAgIGRlcGVuZGVudDogRGVwZW5kZW50LAogICAgfQp9CgovLyB0aGlzIHR5cGUgaXMgbm90IGNvdmFyaWFudAp0eXBlIERlcGVuZGVudDwnYT4gPSBSZWZDZWxsPEJveDxkeW4gZm10OjpEaXNwbGF5ICsgJ2E+PjsKCmZuIG1haW4oKSB7CiAgICBsZXQgY2VsbCA9IFdyb25nVmFyaWFuY2VFeGFtcGxlOjpuZXcoKCksIHxffCBSZWZDZWxsOjpuZXcoQm94OjpuZXcoIiIpKSk7CiAgICBsZXQgcyA9IFN0cmluZzo6ZnJvbSgiSGVsbG8gV29ybGQiKTsKCiAgICAvLyBib3Jyb3dfZGVwZW5kZW50IHVuc291bmQgZHVlIHRvIGluY29ycmVjdGx5IGNoZWNrZWQgdmFyaWFuY2UKICAgICpjZWxsLmJvcnJvd19kZXBlbmRlbnQoKS5ib3Jyb3dfbXV0KCkgPSBCb3g6Om5ldyhzLmFzX3N0cigpKTsKCiAgICAvLyBzIHN0aWxsIGV4aXN0cwogICAgY2VsbC53aXRoX2RlcGVuZGVudCh8XywgZHwgcHJpbnRsbiEoInt9IiwgZC5ib3Jyb3coKSkpOwoKICAgIGRyb3Aocyk7CgogICAgLy8gcyBpcyBnb25lCiAgICBjZWxsLndpdGhfZGVwZW5kZW50KHxfLCBkfCBwcmludGxuISgie30iLCBkLmJvcnJvdygpKSk7Cn0K))
```
Hello World
p���U���
```

A better `_assert_covariance` needs to check whether coercing the `Dependent`s _behind a level of indirection_ is still possible.

Options include coercing `Box<Dependent<'x>>` to `Box<Dependent<'y>>`, or (more close to the actual use-case) coercing `&'y Dependent<'x>` into `&'y Dependent<'y>`.

**Comment by Voultapher** [comment]
Sigh, here we go again. Variance the bane of self-referential abstractions. Thanks for finding the issue and reporting and fixing it!

**Comment by Voultapher** [comment]
With the bug fixed on main, here are the next steps I propose:

- [ ] Release 1.0.2 with fix
- [ ] Yank all prior versions
- [ ] Open a RustSec advisory
- [ ] Open PRs for public dependents still relying on version < 1.0

Please let me know if that aligns with your expectations.

**Comment by steffahn** [comment]
For 0.10, you could also add a point release containing the fix (for example, just the minimal fix changing the single line in the implementation),
or – assuming that it’s supposed to be identical / compatible with 1.0.* versions anyways – release a point release with a [semver trick](https://github.com/dtolnay/semver-trick) style construction (i.e. new version `0.10.3` that simply depends on version `1.0.2` and re-exports the API from it [provided you that that actually works and doesn’t do anything weird like breaking the `$crate::…` macro calls 🤷🏻]; that way, users of `0.10` can also benefit from future patches, without any further work/versions required).

I think it would make sense not to unnecessarily break the `0.10.*` versions completely.

**Comment by Voultapher** [comment]
I'd really like to avoid having to maintain multiple versions, the migration from 0.10 to 1.0 is without any breakage.

**Comment by steffahn** [comment]
I’ve written such a `0.10.3` version myself quickly. As far as I can tell, this works flawlessly (including the `old_rust` feature): https://github.com/steffahn/self_cell/tree/0.10.3

**Comment by Voultapher** [comment]
I find your arguments convincing. I want to publish a `0.10.3` based on your branch, but when I run `cargo t` I get:

```
error[E0464]: multiple candidates for `rlib` dependency `self_cell` found
   --> src/lib.rs:162:9
    |
162 | pub use self_cell::self_cell;
    |         ^^^^^^^^^
    |
    = note: candidate #1: /home/dev/projects/rust/self_cell/target/debug/deps/libself_cell-57d4b17c39f385cc.rlib
    = note: candidate #2: /home/dev/projects/rust/self_cell/target/debug/deps/libself_cell-6694495f354d6a30.rlib
```

**Comment by steffahn** [comment]
Hmm… I don’t think doctests failing is a problem though.

This does exactly follow the pattern of the [semver trick repo](https://github.com/dtolnay/semver-trick/tree/master/semver-trick-0.2.1), as far as I can tell. I think the error makes sense since doctests will add the library itself *and* its dependencies as a dependency, which then means in this case, two dependencies (of the doctest-generated code) with the same name. However, a crate that actually depends on self_cell would only specify one version as a dependency, and it works fine (I tested it locally with a path dependency on the `0.10.3` branch).

**Comment by steffahn** [comment]
It’s interesting though that it does point out the error in the `self_cell` source code here; I’m not familiar with the inner workings of doctests… I’ll check if there’s a workaround (such as renaming the dep) without any negative consequences.

**Comment by steffahn** [comment]
Tested it now; renaming seems to incur no problems, too, so I pushed a commit with the rename.

**Comment by Voultapher** [comment]
Thanks, I've published 0.10.3, and have yanked the affected versions.
